harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,107 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def merge_metadata(all_metadata: List[Dict[str, Any]]) -> Tuple[Dict[str, Any], Dict[str, List]]:
+    """
+      Given a list of metadata, return:
+        - merged metadata
+        - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+    #  Metadata 1        | Metadata 2          | Conflict |  Result
+    # -------------------|---------------------|----------|-------------------------------
+    #  key1: val         | key2: val           | no       | {key1: val, key2: val}
+    #  key1: val         | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: [val]         | no       | {key1: [val, val]}
+    #  key1: [val]       | key1: val           | yes      | n/a
+    #  key1: [val]       | key1: {key3: val}   | yes      | n/a
+    #  key1: {key3: val} | key1: {key4: val}   | no       | key1: {key3: val, key4: val}
+    #  key1: {key3: val} | key1: val           | yes      | n/a
+    #  key1: {key3: val} | key1: [val]         | yes      | n/a
+    """
+    merged: Dict[str, Any] = {}
+    conflicts: Dict[str, List] = {}
+    for rank, rank_metadata in enumerate(all_metadata):
+
+        for _key in rank_metadata:
+            metadata = rank_metadata[_key]
+            if _key not in merged:
+                merged[_key] = metadata
+                conflicts[_key] = [rank]
+            else:
+                if not isinstance(metadata, type(merged[_key])):
```

Comment:
You have this a couple times, where you have:
```python
if condition:
else:
  if condition:
  else:
    if condition:
    else:
```

But these should really be one chain of if/elif with a single else at the end, which would improve readability a lot, because when I see single if/elif chain I already understand the control flow.  When there are nested ifs, I have to read very carefully because I assume that the desired control flow can't be represented as a single if/elif chain.